media_driver/agnostic/common/vp/hal/vphal_renderer.cpp

```diff
@@ -1397,6 +1397,14 @@ MOS_STATUS VphalRenderer::Initialize(
     // We MUST NOT create a writable global memory since it can cause issues
     // in multi-device cases (multiple threads operating on the memory)
     // NOTE: KDLL will release the allocated memory.
+    // NOTE: Need to check kernel binary pointer and bin size firstly,
+    // Because calling malloc(0) will usually returns a valid pointer, and any usage of this pointer excluding free memory is undefined in C/C++.
+    if(pcKernelBin == nullptr || dwKernelBinSize == 0)
+    {
+        eStatus = MOS_STATUS_NULL_POINTER;
+        VPHAL_RENDER_ASSERTMESSAGE("Could not allocate KDLL state with no kernel binary");
+        goto finish;
+    }
```

Comment:
For gen9 with ENABLE_NONFREE_KERNELS=OFF,  pcKernelBin will be nullptr and dwKernelBinSize will be 0.  
If we don't check it  and return error here, it will call MOS_AllocMemory(0) and get a valid pointer. 
Any usage of this pointer  in KernelDll_AllocateStates will be undefined.